#################################################################################
#
#           Gut microbiome structure and function in wild spotted hyenas
#                      
#           Rojas et al 2022. Taxonomic and functional variation of the gut 
#               microbiome over host lifespan in wild spotted hyenas
#
#                         By: Connie Rojas
#                         Created: 18 Nov 2021
#                         Last updated: 20 Jul 2022
#
###############################################################################

## CODE FOR: A) making piechart showing abundances of Archaea, Viruses, 
#                   Eukaryotes and Bacteria
#            B) generating stacked barplots of gut microbiome composition showing
#               abundances of Bacterial Phyla, Orders, and Genera

# Data generated by Kraken2, which outputs the taxonomic profiles of 
#            whole-metagenome shotgun samples using a variety of databases


source(file="scripts/00_background.R"); #load necessary packages and specifications


################################################################################
#             1.  Load taxa abundance table output by Kraken 
#                           and sample metadata
################################################################################

krak=read.table("data/metagenomes/00_Kraken_abundances.txt", 
                 sep="\t", header=T);
load("data/metagenomes/00_metagenomes_metadata.Rdata");

# make a vector of your sample names for later
samples=meta$sampleID;

# subset Kraken 2 data to only Bacteria
krak2=krak[krak$Domain=="Bacteria",];


################################################################################
#             2. Pie chart of domain abundances           
################################################################################

# aggregate data by Domain
kdom=krak[,which(names(krak) 
                 %in% c(samples, "Domain"))];
colnames(kdom)[ncol(kdom)]="taxa";
kdom=aggregate(.~taxa,kdom, sum);

# convert raw counts to proportions
kdom[,-1] <- lapply(kdom[,-1], function(x) (x/sum(x))*100);

# obtain Domain averages across samples
rowMeans(kdom[-1]);

# make data frame of averages
x <-c(88.12, 11.28, sum(0.47,0.11));
labels <-c("Bacteria","Eukaryota","Archaea.Viruses");
piepercent<- round(100*x/sum(x), 1);

# plot
pie(x, labels = labels, main = "",col = c("#bcbddc","#bdbdbd","#fee0d2"))

# save plot manually Save as PDF > "01_domain_pie.pdf"


################################################################################
#             3. Create stacked bar plots of Bacterial Phyla abundances           
################################################################################

# sum up taxa abundances 
kphy=krak2[,which(names(krak2) 
                %in% c(samples, "Phylum"))];
colnames(kphy)[ncol(kphy)]="taxa";
kphy=aggregate(.~taxa,kphy, sum);

# calculate proportions
kphy[,-1] <- lapply(kphy[,-1], function(x) (x/sum(x))*100);
print(colSums(kphy[-1]));

# keep taxa >0.9% relative abundance across samples
kphy$AVG=rowMeans(kphy[,-1]);
kphy=kphy[kphy$AVG>0.9,];
kphy$AVG=NULL;

# denote the rest of phyla as "Other"
newrow=c(NA, 100-colSums(kphy[2:ncol(kphy)])); 
kphy=rbind(kphy, newrow); 
kphy$taxa=as.character(kphy$taxa);
kphy[nrow(kphy),1]="Other";

# melt data frame for ggplot
phybar<-reshape2::melt(kphy, id.vars="taxa",value.name = "abun");
colnames(phybar)[2]="sampleID";
phybar=merge(phybar, meta, by="sampleID");

# set up color palette
phy_col=c('#66c2a5','#fc8d62','#7570b3','#e78ac3','#a6d854','#ffd92f');

# plot
barp1=ggplot(data=phybar, 
              aes(reorder(sampleID,ideal_order),y=abun, fill=taxa))+
  geom_bar(stat = "identity")+
  theme_bw()+ 
  labs(x = "",
       y = "Relative Abundance (%)",
       fill="Bacterial Phylum",
       title="Gut metagenomes")+
  scale_fill_manual(values=phy_col)+
  theme(legend.position="right", 
        legend.text = element_text(size=9),
        legend.title = element_text(size=10, face="bold"),
        legend.key.size = unit(0.8,"line"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        plot.title=element_text(size = 13, face="bold"),
        axis.title.y = element_text(size=10, face="bold"),
        axis.text.y = element_text(size=9),
        axis.ticks.x=element_blank(),
        axis.text.x=element_blank(),
        axis.title.x = element_text(size=9, face="bold")); plot(barp1);


################################################################################
#             4. Create stacked bar plots of Bacterial Order abundances           
################################################################################

# sum up taxa abundances 
kord=krak2[,which(names(krak2) 
                  %in% c(samples, "Order"))];
colnames(kord)[ncol(kord)]="taxa";
kord=aggregate(.~taxa,kord, sum);

# calculate proportions
kord[,-1] <- lapply(kord[,-1], function(x) (x/sum(x))*100);
print(colSums(kord[-1]));

# keep taxa >0.9% relative abundance across samples
kord$AVG=rowMeans(kord[,-1]);
kord=kord[kord$AVG>0.9,];
kord$AVG=NULL;

# denote the rest of phyla as "Other"
newrow=c(NA, 100-colSums(kord[2:ncol(kord)])); 
kord=rbind(kord, newrow); 
kord$taxa=as.character(kord$taxa);
kord[nrow(kord),1]="Other";

# melt data frame for ggplot
ordbar<-reshape2::melt(kord, id.vars="taxa",value.name = "abun");
colnames(ordbar)[2]="sampleID";
ordbar=merge(ordbar, meta, by="sampleID");

# set up color palette
ord_col=c("#8c6bb1","#810f7c","#6baed6","goldenrod2",
         "#08589e","#e7298a","#fc9272","#a50f15",
         "#737373","#66c2a4","black");

# plot
barp2=ggplot(data=ordbar, 
             aes(reorder(sampleID,ideal_order),y=abun, fill=taxa))+
  geom_bar(stat = "identity")+
  theme_bw()+ 
  labs(x = "",
       y = "Relative Abundance (%)",
       fill="Bacterial Order")+
  scale_fill_manual(values=ord_col)+
  theme(legend.position="right", 
        legend.text = element_text(size=9),
        legend.title = element_text(size=10, face="bold"),
        legend.key.size = unit(0.8,"line"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        plot.title=element_text(size = 13, face="bold"),
        axis.title.y = element_text(size=10, face="bold"),
        axis.text.y = element_text(size=9),
        axis.ticks.x=element_blank(),
        axis.text.x=element_blank(),
        axis.title.x = element_text(size=9, face="bold")); plot(barp2);


################################################################################
#             5. Create stacked bar plots of Bacterial Genus abundances           
################################################################################

# sum up taxa abundances 
kgen=krak2[,which(names(krak2) 
                  %in% c(samples, "Genus"))];
colnames(kgen)[ncol(kgen)]="taxa";
kgen=aggregate(.~taxa,kgen, sum);

# calculate proportions
kgen[,-1] <- lapply(kgen[,-1], function(x) (x/sum(x))*100);
print(colSums(kgen[-1]));

# keep taxa >1% relative abundance across samples
kgen$AVG=rowMeans(kgen[,-1]);
kgen=kgen[kgen$AVG>1,];
kgen$AVG=NULL;

# denote the rest of phyla as "Other"
newrow=c(NA, 100-colSums(kgen[2:ncol(kgen)])); 
kgen=rbind(kgen, newrow); 
kgen$taxa=as.character(kgen$taxa);
kgen[nrow(kgen),1]="Other";

# melt data frame for ggplot
genbar<-reshape2::melt(kgen, id.vars="taxa",value.name = "abun");
colnames(genbar)[2]="sampleID";
genbar=merge(genbar, meta, by="sampleID");

# set up color palette
gener_col=c("#771155", "#AA4488", "#CC99BB", "#114477", "#4477AA", "#77AADD", 
         "#117777", "#88CCAA",
         "#777711",  "#bdbdbd",
         "#DDDD77","#774411", "#AA7744", "#DDAA77", "#771122", "#AA4455", "black"); 
# plot
barp3=ggplot(data=genbar, 
             aes(reorder(sampleID,ideal_order),y=abun, fill=taxa))+
  geom_bar(stat = "identity")+
  theme_bw()+ 
  labs(x = "",
       y = "Relative Abundance (%)",
       fill="Bacterial Genus")+
  scale_fill_manual(values=gener_col)+
  theme(legend.position="right", 
        legend.text = element_text(size=9),
        legend.title = element_text(size=10, face="bold"),
        legend.key.size = unit(0.8,"line"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        plot.title=element_text(size = 13, face="bold"),
        axis.title.y = element_text(size=10, face="bold"),
        axis.text.y = element_text(size=9),
        axis.ticks.x=element_blank(),
        axis.text.x=element_blank(),
        axis.title.x = element_text(size=9, face="bold")); plot(barp3);


################################################################################
#             6. save plots!      
###############################################################################

ggsave(filename="01_barplot_phylum_kraken.pdf",
       device="pdf",path="./figures/metagenomes",
       plot=barp1,
       width=5,
       height=3.6,
       units="in",
       dpi=500);

ggsave(filename="01_barplot_order_kraken.pdf",
       device="pdf",path="./figures/metagenomes",
       plot=barp2,
       width=5,
       height=3.6,
       units="in",
       dpi=500);

ggsave(filename="01_barplot_genus_kraken.pdf",
       device="pdf",path="./figures/metagenomes",
       plot=barp3,
       width=6,
       height=3.9,
       units="in",
       dpi=500);

