#################################################################################
#
#           Gut microbiome structure and function in wild spotted hyenas
#                      
#           Rojas et al 2022. Taxonomic and functional variation of the gut 
#               microbiome over host lifespan in wild spotted hyenas
#
#                         By: Connie Rojas
#                         Created: 22 Jun 2022
#                         Last updated:  20 Jul 2022
#
################################################################################

## CODE FOR: A) tidying data generated by Anvio after annotating contigs with
#           Cluster of Orthologous Genes (COG) & Kyoto Enc.Genes Genomes (KEGG)
#           B) appending gene counts to Anvio gene annotations
#           Gene counts calculated using Salmon and are in units of TPM

source(file="scripts/00_background.R"); #load necessary packages and specifications


################################################################################
#             1. Load anvio pathway data and Salmon gene counts from
#       Google Drive folder because they are too large to upload to Git
################################################################################

# download 00_Anvio_GeneAnnotations.txt and 00_Salmon_GeneTPM.Rdata from
# https://tinyurl.com/ycxcmvs3
# and deposit in the data/metagenomes folder for this R project

afun=read.table("data/metagenomes/00_Anvio_GeneAnnotations.txt", 
                sep="\t", header=T, fill=T);
load("data/metagenomes/00_Salmon_GeneTPM.Rdata");
colnames(tpm)[1]="geneID";tpm$geneID=as.character(tpm$geneID);

# remove symbols and punctuation that may give R problems down the line
colnames(afun)[1]="geneID"; 
colnames(afun)[4]="genefunction"
afun$geneID=as.character(afun$geneID);
afun$genefunction=gsub("[[:punct:]]", " ", afun$genefunction);

# only keep annotations that have statistically significant e.values
afun=afun[afun$e_value<=0.05,];


################################################################################
#             2. Tidy and subset KEGG data             
################################################################################

# subset to only KEGG data
kegg=afun[afun$source=="KOfam",];row.names(kegg)=NULL;
kegg=kegg[1:nrow(kegg)-1,];

# extract and tidy kegg annotations
kegg2=kegg[,c(1,4)];
kegg2b=as.data.frame(str_split_fixed(kegg2$genefunction, "EC",2));
kegg2$gf=kegg2b$V1;

kegg2$gf[kegg2$geneID %in% c("6174403", "7739913","7755330")]=
  "ECM component binding autotransporter adhesin";

# append gene counts
kc=merge(tpm, kegg2[,c(1,3)],by="geneID");
kc$geneID=NULL;
keggabun=aggregate(.~gf,kc, sum);  # gf = gene function, kc= df of kegg counts

#save for accessing later
save(keggabun, file="data/metagenomes/02_KEGG_abundances.Rdata");


################################################################################
#             3. Tidy and subset COG data             
################################################################################

# subset to only COG data
cogc=afun[afun$source=="COG20_CATEGORY",];row.names(cogc)=NULL;
cogc=cogc[1:(nrow(cogc)-2),];
cogc=cogc[cogc$e_value<0.05,];

# extract and tidy COG annotations
cogc2=cogc[,c(1,4)];
cogc2b=as.data.frame(str_split_fixed(cogc2$genefunction, "   ",2));
cogc2$gf=cogc2b$V1;

# append gene counts
cc=merge(tpm, cogc2[,c(1,3)],by="geneID");
cc$geneID=NULL;
cogabun=aggregate(.~gf,cc, sum); # gf = gene function, cc= df of cog counts

#save for accessing later
save(cogabun, file="data/metagenomes/02_COG_abundances.Rdata");


